File: 210811_완주하지못한선수.cpp
#include <string>
#include <vector>
#include <iostream>
#include <hash_map>

using namespace std;


int hash_key(string data){
    return data[0] - 'a';
}


string solution(vector<string> participant, vector<string> completion) {
    string answer = "";
    vector<string> hashMap[26];
    
    // save value in hashmap, O(N)    
    for (int i = 0; i < completion.size(); i++) {
        string cur = completion[i];
        int key = hash_key(cur);
        hashMap[key].push_back(cur);
    }

    // find and remove arrived player, O(N * N)
    for (int i = 0; i < participant.size(); i++) {
        int same_flag = 0;
        string cur = participant[i];
        
        int key = hash_key(cur);
        for (int j = 0; j < hashMap[key].size(); j++){
            if (cur.compare(hashMap[key][j]) == 0){ // 일치하는 게 있다면
                hashMap[key].erase(hashMap[key].begin() + j);
                same_flag = 1;
                break;
            }
        }
        if (!same_flag) {
            answer = cur;
            return answer;
        }
    }
	
   return answer;
}

File: Hash/210811_완주하지못한선수.java
package Hash;
import java.util.HashMap;

class Solution {
  public String solution(String[] participant, String[] completion) {
      String answer = "";
      
      HashMap<String, Integer> map = new HashMap<>();

      for (int i = 0; i < completion.length; i++){
        if (map.get(completion[i]) == null){
          map.put(completion[i], 1);
        }
        else {
          map.replace(completion[i], map.get(completion[i]) + 1);
        }
      }
      
      for (int i = 0; i < participant.length; i++){
        Integer value = map.get(participant[i]);
        if (value == null || value == 0){
          answer = participant[i];
          return answer;
        }
        else {
          map.replace(participant[i], map.get(participant[i]) - 1);
        }
      }

      return answer;
  }
}